src/lightning_utilities/core/rank_zero.py

```diff
@@ -9,15 +9,25 @@
 from platform import python_version
 from typing import Any, Callable, Optional, TypeVar, Union
 
-from typing_extensions import ParamSpec
+from typing_extensions import ParamSpec, overload
 
 log = logging.getLogger(__name__)
 
 T = TypeVar("T")
 P = ParamSpec("P")
 
 
+@overload
 def rank_zero_only(fn: Callable[P, T]) -> Callable[P, Optional[T]]:
+    ...
+
+
+@overload
+def rank_zero_only(fn: Callable[P, T], default: T) -> Callable[P, T]:
```

Comment:
this looks strange to have the same func with diff args :thinking:

Reply:
Its only for proper typing: https://docs.python.org/3/library/typing.html#typing.overload

Reply:
@Borda overloading means that the function can have different signatures.